x-pack/plugins/infra/public/utils/use_url_state.ts

```diff
@@ -47,13 +47,24 @@ export const useUrlState = <State>({
   }, [defaultState, decodedState]);
 
   const setState = useCallback(
-    (newState: State | undefined) => {
+    (patch: State | undefined | ((prevState: State) => State)) => {
       if (!history || !history.location) {
         return;
       }
 
       const currentLocation = history.location;
 
+      const newState =
+        patch instanceof Function
+          ? patch(
+              decodeUrlState(
+                decodeRisonUrlState(
+                  getParamFromQueryString(getQueryStringFromLocation(currentLocation), urlStateKey)
+                )
+              ) ?? defaultState
+            )
+          : patch;
+
```

Comment:
Q: Should we have 2 functions in this case (as there are 2 actions - patch the state and set the state) - one to use a state to set it and one to have the patch function as a parameter, I think this way it looks a bit confusing to understand all the cases in this condition.

Reply:
We're basically reproducing the same functionality provided by React's `setState` ([doc](https://react.dev/reference/react/useState#updating-state-based-on-the-previous-state)). Since this hook is supposed to do the same thing in theory, it felt natural that it also provided the same interfaces. How would you name this new second setter function?

Reply:
It's OK to have a function as a parameter in the `setState` it's not what I was trying to replace -  I was thinking about the type of the param and the option to have type function OR type state and separating the logic there so we have something like: 
```
  const patchState = (patch?: (prevState: State) => State) => {...}
  const setNewState = (newState?: State ) => {...}
```
And then use it but it's not an issue I wanted to make this condition easier to read